### PR TITLE
[WIP] Add mTLS https example

### PR DESCRIPTION
**Pull request by theotherjimmy**
This include a test-server written in Rust, for your convenience.

For the time being, this is very much WIP.

### CODE REVIEW

File: samples/net/sockets/mtls_http_get/README.rst [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+Overview
+********
+
+The sockets/http_get sample application for Zephyr implements a simple
+HTTP GET client using a BSD Sockets compatible API and mutual TLS
+authentication. The purpose of this sample is to show how it's possible
+to do mTLS with Zephyr. 
+
+The source code for this sample application can be found at:
+:zephyr_file:`samples/net/sockets/http_get`.
```
Comment (d3zd3z) [inline samples/net/sockets/mtls_http_get/README.rst:15]: This should be fixed for the directory we're in now. Maybe reference where it came from?

File: samples/net/sockets/mtls_http_get/prj.conf [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1"
+CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.2"
+CONFIG_NET_CONFIG_MY_IPV4_GW="192.0.2.2"
+
+# TLS configuration
+CONFIG_MBEDTLS=y
+CONFIG_MBEDTLS_BUILTIN=y
+CONFIG_MBEDTLS_ENABLE_HEAP=y
+CONFIG_MBEDTLS_HEAP_SIZE=40000
+CONFIG_MBEDTLS_SSL_MAX_CONTENT_LEN=7168
```
Comment (d3zd3z) [inline samples/net/sockets/mtls_http_get/prj.conf:36]: This isn't likely to work very well with a generic web server. If we are able to configure the buffer size in our client, though, it should be fine.

File: samples/net/sockets/mtls_http_get/server/client.key [hunk 3]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (d3zd3z) [inline samples/net/sockets/mtls_http_get/server/client.key:1]: I wonder if we should include these, or if we should just leave them out, and instruct people to run the generate script. We tend to get annoying vulnerability reports from people that scan github repos for keys. Realistically, though, I don't have a problem with these keys being in tree.

File: samples/net/sockets/mtls_http_get/server/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    Ok(root)
+}
+
+#[tokio::main]
+async fn main() -> io::Result<()> {
+    let certs = load_certs()?;
+    let mut keys = load_keys()?;
+    let root_store = load_root_store()?;
+
+    let config = rustls::ServerConfig::builder()
```
Comment (d3zd3z) [inline samples/net/sockets/mtls_http_get/server/src/main.rs:43]: At some point, we should try to figure out if the TLS maximum record size can be configured in rustls. We will have problems right now if any request ends up larger than the 7k configured for the maximum record size on the client.

File: samples/net/sockets/mtls_http_get/src/ca_certificate.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __CA_CERTIFICATE_H__
+#define __CA_CERTIFICATE_H__
+
+#define CA_CERTIFICATE_TAG 1
+
+/* By default only certificates in DER format are supported. If you want to use
```
Comment (d3zd3z) [inline samples/net/sockets/mtls_http_get/src/ca_certificate.h:12]: This comment seems in conflict with the PEM format certificates these files point to.

**Review by d3zd3z — CHANGES_REQUESTED** [review]
I think we're close to this being ready, just a few things beyond what I mentioned in code comments. Namely, the commit text should better explain what is happening, and what we're trying to do here. Also, make sure there is a Signed-off-by in the commit text.

**Comment by d3zd3z** [comment]
Also check, there is a stray space at the end of a line in the README.rst file.

**Comment by theotherjimmy** [comment]
The text may be copy-paste from another example, and may not have been edited for this example

**Comment by jukkar** [comment]
The `samples/net/sockets/http_get` supports both http and https protocols. Is there some reason we would need a new app that would do the same stuff?

**Comment by microbuilder** [comment]
> The `samples/net/sockets/http_get` supports both http and https protocols. Is there some reason we would need a new app that would do the same stuff?

The purpose of this sample if to demonstrate **mutual** TLS, where a certificate is also present on the client side which is validated by the server. The supplied Rust project demonstrates what a server that supports this might look like to be able to test both sides of the TLS handshake.

File: samples/net/sockets/mtls_http_get/README.rst [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+- :ref:`networking_with_host`
+
+Building and Running
+********************
+
+Build the Zephyr version of the application like this:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/net/sockets/http_get
```
Comment (microbuilder) [inline samples/net/sockets/mtls_http_get/README.rst:28]:
```suggestion
   :zephyr-app: samples/net/sockets/mtls_http_get
```

File: samples/net/sockets/mtls_http_get/README.rst [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+case no NAT/routing setup is needed).
+
+Enabling TLS support
+=================================
+
+Enable TLS support in the sample by building the project with the
+``overlay-tls.conf`` overlay file enabled, for example, using these commands:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/net/sockets/http_get
```
Comment (microbuilder) [inline samples/net/sockets/mtls_http_get/README.rst:45]:
```suggestion
   :zephyr-app: samples/net/sockets/mtls_http_get
```

File: samples/net/sockets/mtls_http_get/README.rst [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+Socket HTTP GET Example with mTLS
+#################################
+
+Overview
+********
+
+The sockets/http_get sample application for Zephyr implements a simple
+HTTP GET client using a BSD Sockets compatible API and mutual TLS
+authentication. The purpose of this sample is to show how it's possible
+to do mTLS with Zephyr.
```
Comment (d3zd3z) [inline samples/net/sockets/mtls_http_get/README.rst:12]: There is a space at the end of this line.

**Pull request closed without merging**